File: collector/src/main/java/edu/monash/humanise/smartcity/collector/Router.java
package edu.monash.humanise.smartcity.collector;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.Message;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@Slf4j
@Builder
@AllArgsConstructor
public class Router {
    @NonNull
    private final Message<?> message;

    private final boolean isDocker = Objects.equals(System.getenv("TAATTA_DOCKER"), "1");

    public void route() {
        String payload = message.getPayload().toString();
        String topic = message.getHeaders().get("mqtt_receivedTopic", String.class);
        assert topic != null;
        log.info("New message. Topic: {}, payload: {}", topic, payload);

        if (topic.startsWith("athom-smart-plug")) {
            // Topic structure: <hostname>/sensors/<sensor_name>/state
            String[] topicComponents = topic.split("/");
            try {
                JSONObject jsonObject = new JSONObject();
                jsonObject.put("deviceName", topicComponents[0]);
                jsonObject.put("data", payload);
                jsonObject.put("sensor", topicComponents[2]);

                RestTemplate restTemplate = new RestTemplate();
                HttpHeaders headers = new HttpHeaders();
                headers.setContentType(MediaType.APPLICATION_JSON);
                HttpEntity<?> request = new HttpEntity<>(jsonObject.toString(), headers);
                String url = String.format("http://%s:4050/api/payload", isDocker? "athom-smart-plug" : "localhost");
                ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                log.info("Response from {} is : {}",topicComponents[0], response);
            } catch (JSONException e) {
                log.error("Cannot create JSON object for smart plug", e);
                throw new RuntimeException(e);
            }
        } else {
            // from chirpstack
            try {
                JSONObject jsonObject = new JSONObject(payload);
                if (jsonObject.has("data")) {
                    RestTemplate restTemplate = new RestTemplate();

                    MultiValueMap<String, String> headers = new LinkedMultiValueMap<String, String>();
                    Map map = new HashMap<String, String>();
                    map.put("Content-Type", "application/json");
                    headers.setAll(map);

                    HttpEntity<?> request = new HttpEntity<>(payload, headers);
                    String deviceProfile = jsonObject.getString("deviceProfileName");

                    if (deviceProfile.endsWith("rhf1s001")) {
                        String url = String.format("http://%s:4040/api/payload", isDocker? "rhf1s001" : "localhost");
                        ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                        log.info("Response from {} is : {}", deviceProfile, response.toString());
                    } else if (deviceProfile.endsWith("wqm101")) {
                        String url = String.format("http://%s:4042/api/payload", isDocker? "wqm101" : "localhost");
                        ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                        log.info("Response from {} is : {}", deviceProfile, response.toString());
                    } else if (deviceProfile.endsWith("df702")) {
                        String url = String.format("http://%s:4044/api/payload", isDocker? "df702" : "localhost");
                        ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                        log.info("Response from {} is : {}", deviceProfile, response.toString());
                    } else if (deviceProfile.endsWith("tbs220")) {
                        String url = String.format("http://%s:4046/api/payload", isDocker? "tbs220" : "localhost");
                        ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                        log.info("Response from {} is : {}", deviceProfile, response.toString());
                    } else if (deviceProfile.endsWith("pcr2")) {
                        String url = String.format("http://%s:4048/api/payload", isDocker? "pcr2" : "localhost");
                        ResponseEntity<?> response = restTemplate.postForEntity(url, request, String.class);
                        log.info("Response from {} is : {}", deviceProfile, response.toString());
                    } else {
                        log.error("This device type {} not implemented yet.", deviceProfile);
                    }
                } else {
                    log.warn("No data field found.");
                }
            } catch (JSONException e) {
                log.error("Cannot parse JSON", e);
                throw new RuntimeException(e);
            }
        }
    }
}

File: collector/src/main/kotlin/edu/monash/humanise/smartcity/collector/Router.kt
package edu.monash.humanise.smartcity.collector

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import edu.monash.humanise.smartcity.collector.chirpstack.events.ChirpStackUplinkEvent
import edu.monash.humanise.smartcity.collector.payload.EspHomePayload
import edu.monash.humanise.smartcity.collector.payload.LoRaPayload
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.Message
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset

private val logger = KotlinLogging.logger {}

/**
 * Component which routes the message to the appropriate logger modules.
 */
@Component
class Router @Autowired constructor(configLoader: SensorRoutersConfigLoader) {
    /** Sensor config */
    private lateinit var sensorRoutersConfig: SensorRoutersConfig

    /**
     * JSON encoder/decoder for router.
     */
    private val jsonCoder = jacksonObjectMapper()

    // some internal state for logging purposes
    /**
     * Most recent time that [retrySendingData] is called.
     */
    private var previousTime = OffsetDateTime.now(ZoneOffset.UTC)

    init {
        this.sensorRoutersConfig = configLoader.loadConfig()
    }

    /**
     * Routes a message to the appropriate logger modules.
     *
     * @param message A [Message] instance containing MQTT message
     */
    fun route(message: Message<*>) {
        val timestamp = message.headers.timestamp ?: Instant.now().toEpochMilli()
        val payload = message.payload.toString()
        val topic: String = message.headers["mqtt_receivedTopic"] as String
        logger.debug { "New message. Topic: $topic, payload: $payload" }

        if (!topic.startsWith("application")) {
            val topicComponents = topic.split("/")
            val deviceName = topicComponents[0]
            // try matching esphome sensors
            val matchingEspHomeSensor =
                sensorRoutersConfig.espHomeModules.firstOrNull { sensor -> topic.startsWith(sensor.name) }
            if (matchingEspHomeSensor != null) {
                try {
                    val payloadSensor = when (topicComponents[1]) {
                        "status" -> "status"
                        "debug" -> "debug"
                        "switch" -> "switch"
                        else -> topicComponents[2]
                    }
                    val jsonObject = EspHomePayload(timestamp, deviceName, payloadSensor, payload)
                    val results = matchingEspHomeSensor.sendData(jsonCoder.writeValueAsString(jsonObject))
                    results.forEach { result ->
                        result.onSuccess { response -> logger.debug { "Response from ${matchingEspHomeSensor.name} is: $response" } }
                        result.onFailure { error -> logger.debug { "Cannot send data for sensor ${matchingEspHomeSensor.name}: $error" } }
                    }
                } catch (e: JsonProcessingException) {
                    logger.error(e) { "Cannot parse JSON for topic $topic" }
                }
            } else {
                logger.warn { "This device type $deviceName not implemented yet, skipping." }
            }
        } else {
            // handle lora sensors
            try {
                val event: ChirpStackUplinkEvent = jsonCoder.readValue(payload)
                if (event.data != null) {
                    val deviceProfile = event.deviceProfileName
                    val sensor = sensorRoutersConfig.loraModules.firstOrNull { sensorModule ->
                        deviceProfile.endsWith(sensorModule.name)
                    }
                    if (sensor != null) {
                        val jsonObject = LoRaPayload(timestamp, event.deviceName, event.devEUI, event.data)
                        val results = sensor.sendData(jsonCoder.writeValueAsString(jsonObject))
                        results.forEach { result ->
                            result.onSuccess { response -> logger.debug { "Response from ${sensor.name} is: $response" } }
                            result.onFailure { error -> logger.debug { "Cannot send data for sensor ${sensor.name}: $error" } }
                        }
                    } else {
                        logger.warn { "This device type $deviceProfile not implemented yet, skipping." }
                    }
                } else {
                    logger.warn { "Data field not found or null, skipping" }
                }
            } catch (e: JsonProcessingException) {
                logger.error(e) { "Cannot parse JSON for topic $topic" }
            }
        }
    }

    /**
     * Retry any failed requests every minute.
     */
    @Scheduled(cron = "0 * * * * *")
    fun retrySendingData() {
        val now = OffsetDateTime.now(ZoneOffset.UTC)
        logger.info { "Retrying any failed requests from $previousTime to $now" }
        var successCount = 0
        var failCount = 0
        sensorRoutersConfig.espHomeModules.forEach { router ->
            val results = router.retryFailedRequests()
            results.forEach { result ->
                result.onSuccess { response ->
                    logger.debug { "Response from ${router.name} is: $response" }
                    successCount += 1
                }
                result.onFailure { error ->
                    logger.error(error) { "Cannot send data for sensor ${router.name} after a retry" }
                    failCount += 1
                }
            }
        }
        sensorRoutersConfig.loraModules.forEach { router ->
            val results = router.retryFailedRequests()
            results.forEach { result ->
                result.onSuccess { response ->
                    logger.debug { "Response from ${router.name} is: $response" }
                    successCount += 1
                }
                result.onFailure { error ->
                    logger.error(error) { "Cannot send data for sensor ${router.name} after a retry" }
                    failCount += 1
                }
            }
        }

        logger.info { "Number of retried requests: $successCount succeeded, $failCount failed"}
        previousTime = now
    }
}
